Provide more information in PUB_ENVIRONMENT for `pub get/upgrade`

kevmoo opened the pull request:
Should help diagnose and fix https://github.com/dart-lang/pub-dartlang-dart/issues/636

Hixie commented on packages/flutter_tools/lib/src/dart/pub.dart:
```diff
@@ -29,7 +29,7 @@ bool _shouldRunPubGet({ File pubSpecYaml, File dotPackages }) {
 }
 
-Future<Null> pubGet({
+Future<Null> pubGet(String pubContext, {
```
to make the call sites clearer about whether the string represents a command to `pub` or just metadata, can you make this an `@required` named argument?

tvolkert replied:
+1, and maybe just call it `context`?

kevmoo replied:
ack

Hixie commented on packages/flutter_tools/lib/src/dart/pub.dart:
```diff
@@ -84,7 +90,10 @@ typedef String MessageFilter(String message);
+///
+/// [pubContext] provides extra information to package server requests to
+/// understand usage. It must match the regular expression `[a-z][a-z_]*[a-z]`.
+Future<Null> pub(List<String> arguments, String pubContext, {
```
same here

Hixie commented on packages/flutter_tools/lib/src/dart/pub.dart:
```diff
@@ -137,20 +146,22 @@ List<String> _pubCommand(List<String> arguments) {
 
 /// The full environment used when running pub.
-Map<String, String> get _pubEnvironment => <String, String>{
+Map<String, String> _pubEnvironment(String pubContext) => <String, String>{
```
this should probably be renamed to make it clearer that it now does something rather than is something. e.g. `_createPubEnvironment` or something.

kevmoo replied:
ack

Hixie commented:
Test?

Hixie commented:
cc @tvolkert

tvolkert commented on packages/flutter_tools/lib/src/dart/pub.dart:
```diff
@@ -29,7 +29,7 @@ bool _shouldRunPubGet({ File pubSpecYaml, File dotPackages }) {
 }
 
-Future<Null> pubGet({
+Future<Null> pubGet(String pubContext, {
```
Can you please add docs to this method?

kevmoo replied:
done

tvolkert commented on packages/flutter_tools/lib/src/dart/pub.dart:
```diff
@@ -47,18 +47,24 @@ Future<Null> pubGet({
+    final List<String> args = <String>[
+      '--verbosity=warning',
+      command,
+      '--no-precompile'
```
Trailing comma (or just leave the line as it was)

kevmoo replied:
done

kevmoo commented:
> Test?

This only affects headers that are passed to the pub server. Not sure if there is any way to test 'em.

I have verified that come through on our logs, though.

tvolkert commented:
> Test?

+1 -- https://github.com/flutter/flutter/tree/master/packages/flutter_tools/test/dart

tvolkert commented:
lol I should read the most recent comments before commenting (my last comment was not a passive aggressive way to ask for a test).

With a mock `ProcessManager`, you could verify that the right arguments get passed to the pub process.  Your call on whether or not this is worth testing.

kevmoo committed:
Provide more information in PUB_ENVIRONMENT for `pub get/upgrade`

Should help diagnose and fix https://github.com/dart-lang/pub-dartlang-dart/issues/636

kevmoo committed:
Add tests

kevmoo commented:
@tvolkert – PTAL

tvolkert reviewed (approved):
LGTM

kevmoo committed:
oops
